src/Utils.vala

```diff
@@ -427,45 +438,189 @@ public class FeedReader.Utils : GLib.Object {
+		var path = GLib.File.new_for_path(icon_path);
+		try{path.make_directory_with_parents();}catch(GLib.Error e){}
+		string local_filename = icon_path + feed_id.replace("/", "_").replace(".", "_") + ".ico";
+		string etag_filename = icon_path + feed_id.replace("/", "_").replace(".", "_") + ".txt";
 
+		// file already exists
+		if(FileUtils.test(local_filename, GLib.FileTest.EXISTS)
+		|| !FileUtils.test(etag_filename, GLib.FileTest.EXISTS))
+		{
+			var message_head = new Soup.Message("HEAD", icon_url);
 			if(Settings.tweaks().get_boolean("do-not-track"))
-				message_dlIcon.request_headers.append("DNT", "1");
+				message_head.request_headers.append("DNT", "1");
 
-			var status = session.send_message(message_dlIcon);
-			if (status == 200)
+			var status = getSession().send_message(message_head);
+			if(status == 200)
 			{
-				try
+				string etag = message_head.response_headers.get_one("ETag");
+
+				if(etag != null)
 				{
-					FileUtils.set_contents(local_filename, (string)message_dlIcon.response_body.flatten().data, (long)message_dlIcon.response_body.length);
+					if(FileUtils.test(etag_filename, GLib.FileTest.EXISTS)
+					&& etag == getFileContent(etag_filename))
+					{
+						// file exists and is identical to remote file
+						Logger.debug(@"Utils.downloadIcon: etag identical -> icon unchanged $feed_id");
+						return true;
+					}
+					else
+					{
+						Logger.debug(@"Utils.downloadIcon: write etag $etag");
+						try
+						{
+							FileUtils.set_contents(etag_filename, etag);
+						}
+						catch(GLib.FileError e)
+						{
+							Logger.error("Error writing etag: %s".printf(e.message));
+						}
+					}
 				}
-				catch(GLib.FileError e)
+				else
 				{
-					Logger.error("Error writing icon: %s".printf(e.message));
+					Logger.warning(@"no etag for $icon_url");
 				}
-				return true;
 			}
-			Logger.error(@"Error downloading icon for feed: $feed_id");
-			return false;
+			else
+			{
+				Logger.warning(@"HEAD-request failed $icon_url");
+				return false;
+			}
 		}
-		// file already exists
-		return true;
+
+		Logger.debug(@"Utils.downloadIcon: url = $icon_url");
+		var message_dlIcon = new Soup.Message("GET", icon_url);
+		if(Settings.tweaks().get_boolean("do-not-track"))
+			message_dlIcon.request_headers.append("DNT", "1");
+
+		var status = getSession().send_message(message_dlIcon);
+		if(status == 200)
+		{
+			if(FileUtils.test(local_filename, GLib.FileTest.EXISTS))
+			{
+				var remoteMd5 = GLib.Checksum.compute_for_string(GLib.ChecksumType.MD5, (string)message_dlIcon.response_body.flatten().data);
+				var localMd5 = GLib.Checksum.compute_for_string(GLib.ChecksumType.MD5, getFileContent(local_filename));
+				if(remoteMd5 == localMd5)
+				{
+					// file exists and is identical to remote file
+					// we already downloaded it, but there is no need to write to the disc
+					Logger.debug("Utils.downloadIcon: md5 identical -> icon unchanged");
+					return true;
```

Comment:
Reading the file and running MD5 is probably more expensive than just writing the new file to disk.